### PR TITLE
use gopoet and goprotoc

### PR DESCRIPTION
**Pull request by jhump**
This switches the protoc plugin in here to use some other libraries I've worked on that are intended to simplify protoc plugin development in Go and do Go code generation.

WDYT, @dragonsinth?

### CODE REVIEW

File: cmd/protoc-gen-grpchan/protoc-gen-grpchan.go [hunk 1]
```diff
@@ -5,310 +5,122 @@
...
+					AddArg("opts", typeOfCallOptions).
+					SetVariadic(true).
+					AddResult("", names.GoTypeOfResponse(md)).
+					AddResult("", gopoet.ErrorType).
+					Printlnf("out := new(%s)", names.GoTypeForMessage(md.GetOutputType())).
+					Printlnf("err := c.ch.Invoke(ctx, \"/%s/%s\", in, out, opts...)", sd.GetFullyQualifiedName(), md.GetName()).
+					Printlnf("if err != nil {").
+					Printlnf("    return nil, err").
+					Printlnf("}").
+					Printlnf("return out, nil"))
```
Comment (dragonsinth) [inline cmd/protoc-gen-grpchan/protoc-gen-grpchan.go]: This all looks very slick, except for the method body gen.  I wish you could write:

```go
					Printlnf(`
						out := new(%s)
						err := c.ch.Invoke(ctx, "/%s/%s", in, out, opts...)
						if err != nil {
							return nil, err
						}
						return out, nil
					`, names.GoTypeForMessage(md.GetOutputType()), sd.GetFullyQualifiedName(), md.GetName()))

```
and have poet clean up the extra blank lines.

Comment (dragonsinth) [reply cmd/protoc-gen-grpchan/protoc-gen-grpchan.go]: Or perhaps even use go's `text/template` in a very neat way.

Comment (jhump) [reply cmd/protoc-gen-grpchan/protoc-gen-grpchan.go]: I changed it to use templates (which is something that gopoet supports). WDYT?

**Comment by dragonsinth** [comment]
Looks great!  Very readable!

**Pull request merged**